Vorbis decoder can't be instantiated in the content process on Sonoma machines (and earlier)

jyavenard opened the pull request:
#### 2d86811e87f12884205b2f17f132cb919b769ce2
<pre>
Vorbis decoder can&apos;t be instantiated in the content process on Sonoma machines (and earlier)
<a href="https://bugs.webkit.org/show_bug.cgi?id=300083">https://bugs.webkit.org/show_bug.cgi?id=300083</a>
<a href="https://rdar.apple.com/161878608">rdar://161878608</a>

Reviewed by Per Arne Vollan.

The AudioToolbox&apos;s AudioComponentFetchServerRegistrations appears non-functional
on Sonoma and earlier, which is what is relied on to enable the opus and vorbis
decoder. We re-enable the old code to register the audio decoder right
before setting up the sandbox in the WebProcess.

The tests had been marked as failing and so the problem had been missed.
This became more problematic as we are trying to move the MediaPlayerPrivate
back into the WebProcess and audio decoding was non functional.

Re-enabled tests.
* LayoutTests/platform/ios/TestExpectations: Re-enable tests.
* LayoutTests/platform/mac-wk2/TestExpectations: re-enable tests.
* Source/WTF/wtf/PlatformEnableCocoa.h: Add ENABLE_AUDIO_DECODER_REGISTRATION
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::initializeSandbox): Register Opus and Vorbis decoder just before setting the sandbox.

Canonical link: <a href="https://commits.webkit.org/300936@main">https://commits.webkit.org/300936@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8207257e94c9d0c6b269f6025991eb4bb41af9f9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124365 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44045 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34775 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131205 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76405 "Built successfully") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44791 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52644 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94604 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/62760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127319 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35699 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111236 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75190 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e6b91272-4404-4c1a-889c-beee854e7ba3) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34636 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29396 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74681 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/116496 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105444 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29618 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133870 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/122893 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51263 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39119 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103083 "Passed tests") | | 
| | [❌ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51662 "Failed to checkout and rebase branch from PR 51736") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107454 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102878 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26182 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26494 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48192 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51130 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56915 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/153990 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50559 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39131 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53917 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52233 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1039,6 +1039,10 @@
 #endif
 
+#if !defined(HAVE_REQUIRE_AUDIO_DECODER_REGISTRATION) && (PLATFORM(MAC) && __MAC_OS_X_VERSION_MIN_REQUIRED < 150000)
+#define HAVE_REQUIRE_AUDIO_DECODER_REGISTRATION 1
```
I wonder if it would be an option to put this in PlatformEnableCocoa.h as `ENABLE_AUDIO_DECODER_REGISTRATION`?

jyavenard replied:
done

pvollan replied:
Looking closer, I see that we call these functions before entering the sandbox. Is there any way to avoid that?

jyavenard committed:
Vorbis decoder can't be instantiated in the content process on Sonoma machines (and earlier)
https://bugs.webkit.org/show_bug.cgi?id=300083
rdar://161878608

Reviewed by Per Arne Vollan.

The AudioToolbox's AudioComponentFetchServerRegistrations appears non-functional
on Sonoma and earlier, which is what is relied on to enable the opus and vorbis
decoder. We re-enable the old code to register the audio decoder right
before setting up the sandbox in the WebProcess.

The tests had been marked as failing and so the problem had been missed.
This became more problematic as we are trying to move the MediaPlayerPrivate
back into the WebProcess and audio decoding was non functional.

Re-enabled tests.
* LayoutTests/platform/ios/TestExpectations: Re-enable tests.
* LayoutTests/platform/mac-wk2/TestExpectations: re-enable tests.
* Source/WTF/wtf/PlatformEnableCocoa.h: Add ENABLE_AUDIO_DECODER_REGISTRATION
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::initializeSandbox): Register Opus and Vorbis decoder just before setting the sandbox.

Canonical link: https://commits.webkit.org/300936@main
